src/ZuluSCSI_disk.cpp

```diff
@@ -500,21 +540,160 @@ static void scsiDiskLoadConfig(int target_idx, const char *section)
+    {
+        ini_gets(section, "ImgDir", "", tmp, sizeof(tmp), CONFIGFILE);
+        if (tmp[0])
+        {
+            logmsg("-- SCSI", target_idx, " using image directory \'", tmp, "'");
+            img.image_directory = true;
+        }
+    }
 }
 
-// Check if image file name is overridden in config
-bool scsiDiskGetImageNameFromConfig(image_config_t &img, char *buf, size_t buflen)
+int scsiDiskGetNextImageName(image_config_t &img, char *buf, size_t buflen)
 {
     int target_idx = img.scsiId & 7;
 
     char section[6] = "SCSI0";
     section[4] = '0' + target_idx;
 
-    char key[5] = "IMG0";
-    key[3] = '0' + img.image_index;
+    // sanity check: is provided buffer is long enough to store a filename?
+    if (buflen < MAX_FILE_PATH)
+    {
+        panic("scsiDiskGetNextImageName called with illegal buflen");
+    }
+
+    if (img.image_directory)
+    {
+        // is this needed to prevent an inadvertent callback during all these opens?
+        platform_set_sd_callback(NULL, NULL);
+
+        // image directory was found during startup
+        FsFile dir;
+        char key[] = "ImgDir";
+        ini_gets(section, key, "", buf, buflen, CONFIGFILE);
+        if (buf[0] != '\0')
+        {
+            if (!dir.open(buf))
+            {
+                logmsg("ImgDir '", buf, "' couldn't be opened.");
+            }
+            if (!dir.isDir())
+            {
+                logmsg("ImgDir '", buf, "' is not a directory.");
+                dir.close();
+                return false;
+            }
+        }
+        else
+        {
+            // If image_directory set but ImageDir is not, could be used to
+            // indicate an image directory configured via folder structure.
+            // Not implemented, so treat this as equivalent to missing ImageDir
+            return false;
+        }
+
+        // Find the filename with the lowest lexical order _after_ the
+        // currently selected file. If there is none, or if there is no
+        // current file, use the lowest filename encountered.
+        char first_name[MAX_FILE_PATH] = {'\0'};
+        char candidate_name[MAX_FILE_PATH] = {'\0'};
+        FsFile file;
+        while (file.openNext(&dir, O_RDONLY))
+        {
+            if (file.isDir()) continue;
+            if (!file.getName(buf, MAX_FILE_PATH))
+            {
+                logmsg("Image directory had invalid file for ID", target_idx);
+                continue;
+            }
+            if (!scsiDiskFilenameValid(buf)) continue;
+
+            // keep track of the first item to allow wrapping
+            // without having to iterate again
+            if (first_name[0] == '\0' || strcmp(buf, first_name) < 0)
+            {
+                strncpy(first_name, buf, sizeof(first_name));
+            }
+
+            // discard if no selected name, or if candidate is before (or is) selected
+            if (img.current_image[0] == '\0' || strcmp(buf, img.current_image) <= 0) continue;
```

Comment:
Might consider `strcasecmp()` here. That way the order would be more natural "A1, a2, B1, b2" instead of "A1, B1, a2, b2".

Reply:
That is definitely more intuitive!